nodejs/aws-infra/experimental/clusterAutoScaling.ts

```diff
@@ -0,0 +1,566 @@
+    const mountPath = args.fileSystem ? args.fileSystem.mountPath : undefined;
+
+    return pulumi.all([cluster.instance.id, cloudFormationStackName, fileSystemId, mountPath])
+                 .apply(([clusterId, cloudFormationStackName, fileSystemId, mountPath]) => {
+        let fileSystemRuncmdBlock = "";
+
+        if (fileSystemId) {
+            // This string must be indented exactly as much as the block of commands it's inserted
+            // into below!
+            mountPath = mountPath || "/mnt/efs";
+
+            // tslint:disable max-line-length
+            fileSystemRuncmdBlock = `
+                # Create EFS mount path
+                mkdir ${mountPath}
+                chown ec2-user:ec2-user ${mountPath}
+                # Create environment variables
+                EFS_FILE_SYSTEM_ID=${fileSystemId}
+                DIR_SRC=$AWS_AVAILABILITY_ZONE.$EFS_FILE_SYSTEM_ID.efs.$AWS_REGION.amazonaws.com
+                DIR_TGT=${mountPath}
+                # Update /etc/fstab with the new NFS mount
+                cp -p /etc/fstab /etc/fstab.back-$(date +%F)
+                echo -e \"$DIR_SRC:/ $DIR_TGT nfs4 nfsvers=4.1,rsize=1048576,wsize=1048576,hard,timeo=600,retrans=2 0 0\" | tee -a /etc/fstab
+                mount -a -t nfs4
+                # Restart Docker
+                docker ps
+                service docker stop
+                service docker start
+            `;
+        }
+
+        return `#cloud-config
+        repo_upgrade_exclude:
+            - kernel*
+        packages:
+            - aws-cfn-bootstrap
+            - aws-cli
+            - nfs-utils
+        mounts:
+            - ['/dev/xvdb', 'none', 'swap', 'sw', '0', '0']
+        bootcmd:
+            - mkswap /dev/xvdb
+            - swapon /dev/xvdb
+            - echo ECS_CLUSTER='${clusterId}' >> /etc/ecs/ecs.config
+            - echo ECS_ENGINE_AUTH_TYPE=docker >> /etc/ecs/ecs.config
+        runcmd:
+            # Set and use variables in the same command, since it's not obvious if
+            # different commands will run in different shells.
+            - |
+                # Knock one letter off of availability zone to get region.
+                AWS_AVAILABILITY_ZONE=$(curl -s 169.254.169.254/2016-09-02/meta-data/placement/availability-zone)
+                AWS_REGION=$(echo $AWS_AVAILABILITY_ZONE | sed 's/.$//')
+
+                ${fileSystemRuncmdBlock}
+
+                # Disable container access to EC2 metadata instance
+                # See http://docs.aws.amazon.com/AmazonECS/latest/developerguide/instance_IAM_role.html
+                iptables --insert FORWARD 1 --in-interface docker+ --destination 169.254.169.254/32 --jump DROP
+                service iptables save
+
+                /opt/aws/bin/cfn-signal \
+                    --region "\${AWS_REGION}" \
+                    --stack "${cloudFormationStackName}" \
+                    --resource Instances
+        `;
+    });
```

Comment:
todo.  is there a sensible way we can provide all this code, wihle still allowing customers to extend this sensibly?